src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -333,61 +244,247 @@ func (a *StorageMinerActorCode_I) _onSuccessfulPoSt(rt Runtime, postSubmission p
 	// pledgeCollateralSatisfied
 
+	// Now that all is done update pointer to last response
 	h, st = a.State(rt)
-	st.ChallengeStatus().Impl().OnChallengeResponse(rt.CurrEpoch())
+	st.ChallengeStatus().Impl().OnPoStSuccess(rt.CurrEpoch())
 	st._processStagedCommittedSectors(rt)
 	UpdateRelease(rt, h, st)
 
 	return rt.SuccessReturn()
 
 }
 
-func (a *StorageMinerActorCode_I) SubmitSurprisePoSt(rt Runtime, postSubmission poster.PoStSubmission) InvocOutput {
+// called by verifier to update miner state on successful surprise post
+func (a *StorageMinerActorCode_I) SubmitSurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) InvocOutput {
 	TODO() // TODO: validate caller
 
+	// Verify correct PoSt Submission
+	isPoStVerified := a._verifySurprisePoSt(rt, onChainInfo)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitSurprisePoSt again
+		// TODO: determine proper error here and error-handling machinery
+		rt.Abort("TODO")
+	}
+
+	return a._onSuccessfulPoSt(rt, onChainInfo)
+
+}
+
+func (a *StorageMinerActorCode_I) _verifySurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
+
+	// 1. Check that the miner in question is currently being challenged
 	if !a._isChallenged(rt) {
 		// TODO: determine proper error here and error-handling machinery
 		rt.Abort("cannot SubmitSurprisePoSt when not challenged")
 	}
 
-	// Verify correct PoSt Submission
-	isPoStVerified := a._verifyPoStSubmission(rt, postSubmission)
-	if !isPoStVerified {
-		// no state transition, just error out and miner should submitSurprisePoSt again
+	// to pull in from constants
+	PROVING_PERIOD := block.ChainEpoch(0)
+
+	// 2. Check that the challenge has not expired
+	// Check that miner can still submit (i.e. that the challenge window has not passed)
+	// This will prevent miner from submitting a Surprise PoSt past the challenge period
+	_, st := a.State(rt)
+	if rt.CurrEpoch() < st.ChallengeStatus().LastChallengeEpoch()+PROVING_PERIOD {
+		rt.Abort("cannot SubmitSurprisePoSt late")
+	}
+
+	// A proof must be a valid snark proof with the correct public inputs
+
+	// 3. Get appropriate randomness
+	surpriseRand := rt.Randomness(onChainInfo.PoStEpoch(), 0) //(st.ChallengeStatus().LastChallengeEpoch(), 0)
+
+	// 4. Get public inputs
+	info := st.Info()
+	sectorSize := info.SectorSize()
+
+	postCfg := sector.PoStCfg_I{
+		Type_:        sector.PoStType_SurprisePoSt,
+		SectorSize_:  sectorSize,
+		WindowCount_: info.WindowCount(),
+		Partitions_:  info.SurprisePoStPartitions(),
+	}
+
+	pvInfo := sector.PoStVerifyInfo_I{
+		OnChain_:    onChainInfo,
+		PoStCfg_:    &postCfg,
+		Randomness_: surpriseRand,
+	}
+
+	sdr := filproofs.WinSDRParams(&filproofs.SDRCfg_I{SurprisePoStCfg_: &postCfg})
+
+	// 5. Verify the PoSt Proof
+	isPoStVerified := sdr.VerifySurprisePoSt(&pvInfo)
+	return isPoStVerified
+}
+
+func (a *StorageMinerActorCode_I) VerifyElectionPoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
```

Comment:
@sternhenri do I understand correctly that this method is read-only and will never actually be invoked "on chain", but as part of block validation. I.e. this same code could exist outside the actor itself and everything would be fine.

Reply:
Absolutely right. Given its use of a lot of things from runtime I decided to add it here, but had started implementing it in the `StorageMiningSubsystem` initially. Given your earlier q about querying chainstate, I could move it there. IN general I feel like a lot of read-only stuff is in our actor code which is a mistake. Actor state should be made equally accessible outside the VM.

What do you think?

Reply:
I just moved it there @anorth. I think it's cleaner to leave readonly methods in the subsystem (where the miner should call them ahead of calling the actor methods anyways)